presto-common/src/main/java/com/facebook/presto/common/type/SqlTimestamp.java

```diff
@@ -20,42 +20,57 @@
 import java.time.format.DateTimeFormatter;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.concurrent.TimeUnit;
 
 import static com.facebook.presto.common.type.TimeZoneKey.UTC_KEY;
+import static java.lang.Math.floorDiv;
+import static java.lang.Math.floorMod;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MICROSECONDS;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 public final class SqlTimestamp
 {
+    private static final long MICROS_PER_SECOND = 1_000_000;
+    private static final long NANOS_PER_MICROS = 1_000;
+
     // This needs to be Locale-independent, Java Time's DateTimeFormatter compatible and should never change, as it defines the external API data format.
-    public static final String JSON_FORMAT = "uuuu-MM-dd HH:mm:ss.SSS";
-    public static final DateTimeFormatter JSON_FORMATTER = DateTimeFormatter.ofPattern(JSON_FORMAT);
+    private static final String JSON_MILLIS_FORMAT = "uuuu-MM-dd HH:mm:ss.SSS";
+    public static final DateTimeFormatter JSON_MILLIS_FORMATTER = DateTimeFormatter.ofPattern(JSON_MILLIS_FORMAT);
+
+    private static final String JSON_MICROS_FORMAT = "uuuu-MM-dd HH:mm:ss.SSSSSS";
+    public static final DateTimeFormatter JSON_MICROS_FORMATTER = DateTimeFormatter.ofPattern(JSON_MICROS_FORMAT);
 
-    private final long millis;
+    private final long value;
     private final Optional<TimeZoneKey> sessionTimeZoneKey;
+    private final TimeUnit precision;
```

Comment:
TimeUnit comes from juc package and usually has a specific meaning. Wondering if ChronoUnit or a similar class from java.time is a better fit here.

Reply:
I am not up to date with Java packages and TimeUnit was from Java 6 days. If you think, something would be better, I can check them. I will check ChronoUnit and java.time packages in the mean time.

the precision is not exposed, so it would be easier to change if required later.

Reply:
I will send out a different PR to change this to ChronoUnit from TimeUnit, this is internal detail and does not change the API.